kernel/userspace.c

```diff
@@ -94,6 +98,185 @@ struct perm_ctx {
 	struct k_thread *parent;
 };
 
+#ifdef CONFIG_DYNAMIC_OBJECTS
+struct dyn_obj {
+	struct _k_object kobj;
+	struct rbnode node; /* must be immediately before data member */
+	u8_t data[]; /* The object itself */
+};
+
+struct visit_ctx {
+	_wordlist_cb_func_t func;
+	void *original_context;
+};
+
+extern struct _k_object *_k_object_gperf_find(void *obj);
+extern void _k_object_gperf_wordlist_foreach(_wordlist_cb_func_t func,
+					     void *context);
+
+static int node_lessthan(struct rbnode *a, struct rbnode *b);
+
+static struct rbtree obj_rb_tree = {
+	.lessthan_fn = node_lessthan
+};
+
+/* TODO: incorporate auto-gen with Leandro's patch */
+static size_t obj_size_get(enum k_objects otype)
+{
+	switch (otype) {
+	case K_OBJ_ALERT:
+		return sizeof(struct k_alert);
+	case K_OBJ_MSGQ:
+		return sizeof(struct k_msgq);
+	case K_OBJ_MUTEX:
+		return sizeof(struct k_mutex);
+	case K_OBJ_PIPE:
+		return sizeof(struct k_pipe);
+	case K_OBJ_SEM:
+		return sizeof(struct k_sem);
+	case K_OBJ_STACK:
+		return sizeof(struct k_stack);
+	case K_OBJ_THREAD:
+		return sizeof(struct k_thread);
+	case K_OBJ_TIMER:
+		return sizeof(struct k_timer);
+	default:
+		return sizeof(struct device);
+	}
+}
+
+static int node_lessthan(struct rbnode *a, struct rbnode *b)
+{
+	return a < b;
+}
+
+static inline struct dyn_obj *node_to_dyn_obj(struct rbnode *node)
+{
+	return CONTAINER_OF(node, struct dyn_obj, node);
+}
+
+static struct dyn_obj *dyn_object_find(void *obj)
+{
+	struct rbnode *node;
+	struct dyn_obj *ret;
+	int key;
+
+	/* For any dynamically allocated kernel object, the object
+	 * pointer is just a member of the conatining struct dyn_obj,
+	 * so just a little arithmetic is necessary to locate the
+	 * corresponding struct rbnode
+	 */
+	node = (struct rbnode *)((char *)obj - sizeof(struct rbnode));
+
+	key = irq_lock();
+	if (rb_contains(&obj_rb_tree, node)) {
+		ret = node_to_dyn_obj(node);
+	} else {
+		ret = NULL;
+	}
+	irq_unlock(key);
+
+	return ret;
+}
+
+void *k_object_alloc(enum k_objects otype)
+{
+	struct dyn_obj *dyn_obj;
+	int key;
+
+	/* Stacks are not supported, we don't yet have mem pool APIs
+	 * to request memory that is aligned
+	 */
+	__ASSERT(otype > K_OBJ_ANY && otype < K_OBJ_LAST &&
+		 otype != K_OBJ__THREAD_STACK_ELEMENT,
+		 "bad object type requested");
+
+	dyn_obj = k_malloc(sizeof(*dyn_obj) + obj_size_get(otype));
```

Comment:
You don't clear the data area here, which isn't incorrect.  I do worry though that almost all kernel objects right now are being allocated in bss, and probably will have bugs if fed unzeroed memory.  Might not be a terrible idea to clear the thing, though it does have performance implications.

Alternatively a stress test where you create and fuzz the initial state of a bunch of kernel objects would be useful to write...